export list.sort

S-YOU opened the pull request:
I think basic list.sort can be exported, most tests cases are already there.

Note that, cmd, key, and reversed options are not yet implemented.

meadori commented on runtime/list_test.go:
```diff
@@ -408,6 +408,8 @@ func TestListSort(t *testing.T) {
 		{args: wrapArgs(newTestList(1, 2, 0, 3)), want: newTestRange(4).ToObject()},
 		{args: wrapArgs(newTestRange(100)), want: newTestRange(100).ToObject()},
+		{args: wrapArgs(1), wantExc: mustCreateException(TypeErrorType, "'TestListSort' requires a 'list' object but received a \"int\"")},
+		{args: wrapArgs(newTestList("foo", "bar"), 2), wantExc: mustCreateException(TypeErrorType, "'TestListSort' requires 1 arguments")},
```
Isn't this really testing the test and not the sort implementation?  Not sure that is necessary.

S-YOU replied:
I think I did that because of coverage missing on 338~340.
The name 'TestListSort' is just because of fun is created inline there.

meadori replied:
The first test that was added to exercise 338-340 is fine.  It is this test I am wondering about:

```
{args: wrapArgs(1), wantExc: mustCreateException(TypeErrorType, "'TestListSort' requires a 'list' object but received a \"int\"")},
```

Doesn't that just test that the `fun` testing function throws an exception? i.e. it doesn't test any implementation code, just test code.

S-YOU replied:
Ah, you are right, I missed there is checkFunctionArgs inside that test function, removing last one...

S-YOU replied:
Removed both, Thank you very much @meadori.

trotterdylan commented on runtime/list_test.go:
```diff
@@ -392,8 +392,8 @@ func TestListSetItem(t *testing.T) {
 func TestListSort(t *testing.T) {
-	fun := newBuiltinFunction("TestListSetItem", func(f *Frame, args Args, _ KWArgs) (*Object, *BaseException) {
-		if raised := checkFunctionArgs(f, "TestListSetItem", args, ListType); raised != nil {
+	fun := newBuiltinFunction("TestListSort", func(f *Frame, args Args, _ KWArgs) (*Object, *BaseException) {
```
To get some coverage over the listSort() code, you could make this test call list.sort() rather than *List.Sort(). Then you could test the arg validating logic as well.

trotterdylan commented on runtime/list.go:
```diff
@@ -334,10 +334,20 @@ func listSetItem(f *Frame, o, key, value *Object) *BaseException {
 	return f.RaiseType(TypeErrorType, fmt.Sprintf("list indices must be integers, not %s", key.Type().Name()))
 }
 
+func listSort(f *Frame, args Args, _ KWArgs) (*Object, *BaseException) {
```
As mentioned in the sorted() PR, list.sort() is a little more complicated than this:

`list.sort(cmp=None, key=None, reverse=False)`

A TODO is fine for now.

S-YOU replied:
Added a TODO

S-YOU commented:
Added
- python side tests
- TODO about cmp=None, key=None, reverse=False)
- and reworked testing code to pass through arguments

Please take a look, @trotterdylan, @meadori

trotterdylan commented on runtime/list_test.go:
```diff
@@ -392,22 +392,20 @@ func TestListSetItem(t *testing.T) {
-		l := toListUnsafe(args[0])
-		if raised := l.Sort(f); raised != nil {
+	fun := newBuiltinFunction("TestListSort", func(f *Frame, args Args, _ KWArgs) (*Object, *BaseException) {
+		if _, raised := listSort(f, args, KWArgs{}); raised != nil {
```
Prefer nil to KWArgs{}

trotterdylan replied:
I prefer to get the list.sort attribute and then Call(). Using listSort() directly is kind of like a back door. You can fetch it once above `fun` and then use it on each test:

```
sort := mustNotRaise(GetAttr(NewRootFrame(), ListType.ToObject(), NewStr("sort"), nil))
fun := ... {
  if _, raised := sort.Call(f, args, nil); raised != nil {
  ...
```

S-YOU replied:
fixed this.

S-YOU replied:
Updated to reflect that. 

One question, so we are testing 
```
func (l *List) Sort(f *Frame) (raised *BaseException) {
```

Instead of 
```
func listSort(f *Frame, args Args, _ KWArgs) (*Object, *BaseException) {
```

Correct?

trotterdylan replied:
You're indirectly testing both. `GetAttr(f, ListType.ToObject(), NewStr("sort"), nil)` is the equivalent of `getattr(list, 'sort')`. By calling sort this way, you're invoking listSort, which is in turn calling `func (l *List) Sort()`.

S-YOU replied:
Thanks for the explanations, I got it now.

trotterdylan commented:
This is great! Thanks for getting this done.
